File: src/components/ClipboardManager/utils/sensitive-data/detectors/detectPassword.ts
import type { SensitiveMatch } from '../sensitiveData.types'

const PASSWORD_PATTERNS: Array<{ regex: RegExp; label: string }> = [
  { regex: /(?:password|passwd|pwd)\s*[:=]\s*['"]?[^\s'"]{4,}['"]?/gi, label: 'Password' },
  { regex: /(?:secret|token|auth)\s*[:=]\s*['"]?[^\s'"]{8,}['"]?/gi, label: 'Secret/Token' },
]

export function detectPassword(text: string): SensitiveMatch[] {
  const matches: SensitiveMatch[] = []

  for (const { regex, label } of PASSWORD_PATTERNS) {
    regex.lastIndex = 0
    let match: RegExpExecArray | null
    while ((match = regex.exec(text)) !== null) {
      matches.push({
        type: 'password',
        label,
        level: 'critical',
        start: match.index,
        end: match.index + match[0].length,
      })
    }
  }

  return matches
}

File: src-tauri/src/commands/clipboard/analysis/sensitive/detect_password.rs
use super::super::utf16_offset::byte_to_utf16_offset;
use crate::types::SensitivityMatch;
use regex::Regex;
use std::sync::OnceLock;

/// Detects inline password / secret assignments. Faithful port of the frontend
/// `detectPassword` sensitive-data detector. Matches are `critical`.
pub fn detect_password(text: &str) -> Vec<SensitivityMatch> {
    static PATTERNS: OnceLock<Vec<(Regex, &'static str)>> = OnceLock::new();
    let patterns = PATTERNS.get_or_init(|| {
        vec![
            (
                Regex::new(r#"(?i)(?:password|passwd|pwd)\s*[:=]\s*['"]?[^\s'"]{4,}['"]?"#)
                    .expect("re"),
                "Password",
            ),
            (
                Regex::new(r#"(?i)(?:secret|token|auth)\s*[:=]\s*['"]?[^\s'"]{8,}['"]?"#)
                    .expect("re"),
                "Secret/Token",
            ),
        ]
    });

    let mut matches = Vec::new();
    for (regex, label) in patterns {
        for m in regex.find_iter(text) {
            matches.push(SensitivityMatch {
                kind: "password".to_string(),
                label: (*label).to_string(),
                level: "critical".to_string(),
                start: byte_to_utf16_offset(text, m.start()),
                end: byte_to_utf16_offset(text, m.end()),
            });
        }
    }
    matches
}
